File: parser/props/DNALandmarkBuilding.py
import struct

import DNANode


class DNALandmarkBuilding(DNANode.DNANode):
    PROP_CODE = 13

    def __init__(self, name):
        DNANode.DNANode.__init__(self, name)

        self.code = ''
        self.wallColor = (1, 1, 1, 1)
        self.title = ''
        self.article = ''
        self.buildingType = ''

    def setCode(self, code):
        self.code = code

    def getCode(self):
        return self.code

    def setWallColor(self, color):
        self.wallColor = color

    def getWallColor(self):
        return self.wallColor

    def setTitle(self, title):
        self.title = title

    def getTitle(self):
        return self.title

    def setArticle(self, article):
        self.article = article

    def getArticle(self):
        return self.article

    def setBuildingType(self, buildingType):
        self.buildingType = buildingType

    def getBuildingType(self):
        return self.buildingType

    def debug(self, message):
        if self.verbose:
            print 'DNALandmarkBuilding:', message

    def traverse(self, recursive=True, verbose=False):
        data = DNANode.DNANode.traverse(self, recursive=False, verbose=verbose)

        code = self.getCode()
        data += struct.pack('>B', len(code))  # Code length
        self.debug('packing... code length: {0}'.format(len(code)))
        data += code
        self.debug('packing... code: {0}'.format(code))

        for component in self.getWallColor():
            data += struct.pack('f', component)  # Wall color
            self.debug('packing... wall color: {0}'.format(component))

        title = self.getTitle()
        if not title:
            data += struct.pack('>B', 0)  # Title length
            self.debug('skipping... title length')
            self.debug('skipping... title')
        else:
            data += struct.pack('>B', len(title))  # Title length
            self.debug('packing... title length: {0}'.format(len(title)))
            data += title  # Title
            self.debug('packing... title: {0}'.format(title))

        article = self.getArticle()
        if not article:
            data += struct.pack('>B', 0)  # Article length
            self.debug('skipping... article length')
            self.debug('skipping... article')
        else:
            data += struct.pack('>B', len(article))  # Article length
            self.debug('packing... article length: {0}'.format(len(article)))
            data += article  # Article
            self.debug('packing... article: {0}'.format(article))

        buildingType = self.getBuildingType()
        if self.getBuildingType():
            data += struct.pack('>B', 0)  # Building type length
            self.debug('skipping... building type length: {0}'.format(len(buildingType)))
            self.debug('skipping... building type: {0}'.format(buildingType))
        else:
            data += struct.pack('>B', len(buildingType))  # Building type length
            self.debug('packing... building type length: {0}'.format(len(buildingType)))
            data += buildingType  # Building type
            self.debug('packing... building type: {0}'.format(buildingType))

        if recursive:
            for child in self.children:
                data += child.traverse(recursive=True, verbose=verbose)

        return data

File: reader/DNALandmarkBuilding.cpp
#include "DNALandmarkBuilding.h"

DNALandmarkBuilding::DNALandmarkBuilding(string name) : DNANode(name), m_wall_color(new LVector4f(1)) {}
DNALandmarkBuilding::~DNALandmarkBuilding(void) {}

void DNALandmarkBuilding::make_from_dgi(DatagramIterator& dgi, DNAStorage* store)
{
	DNANode::make_from_dgi(dgi, store);
	m_code = dgi_extract_string8(dgi);
	m_wall_color = dgi_extract_color(dgi);
	m_title = dgi_extract_string8(dgi);
	m_article = dgi_extract_string8(dgi);
	m_building_type = dgi_extract_string8(dgi);
};

void DNALandmarkBuilding::traverse(NodePath& np, DNAStorage* store)
{
	NodePath* result = store->find_node(m_code);
	if (result->is_empty())
	{
		stringstream ss;
		ss << "code " << m_code << " not found in storage!";
		cerr << "DNALandmarkBuilding(fatal): " << ss.str() << endl;
		throw ss.str();
	}

	NodePath _np = result->copy_to(np);
	_np.set_name(m_name);

	_np.set_pos(*m_pos);
	_np.set_hpr(*m_hpr);
	_np.set_scale(*m_scale);

	setup_suit_building_origin(np, _np);

	for (comp_vec_t::iterator child = m_children.begin(); child != m_children.end(); child++)
	{
        (*child)->traverse(_np, store);
	};

	_np.flatten_strong();
};

void DNALandmarkBuilding::setup_suit_building_origin(NodePath a, NodePath b)
{
	locale loc;
	bool is_tb = (m_name.substr(0, 2) == "tb");
	bool is_digit = (string(1, m_name.at(2)).find_first_not_of("0123456789") == std::string::npos);
	bool has_colon = (m_name.find(':') != string::npos);
	if (is_tb && is_digit && has_colon)
	{
		stringstream ss;
		ss << "s" << m_name.substr(1);
		string name = ss.str();

		NodePath node = b.find("**/*suit_building_origin");
		if (node.is_empty())
		{
			cout << "DNALandmarkBuilding " << name << " did not find **/*suit_building_origin" << endl;
			node = a.attach_new_node(name);
			node.set_pos(*m_pos);
			node.set_hpr(*m_hpr);
			node.set_scale(*m_scale);
		}

		else
		{
			node.wrt_reparent_to(a);
			node.set_name(name);
		}
	}

	else
	{
		cout << "DNALandmarkBuilding " << m_name << " does not match needed conditions to setup suit origin" << endl;
	};
};

void DNALandmarkBuilding::set_article(const string& article)
{
	m_article = article;
};

string DNALandmarkBuilding::get_article()
{
	return m_article;
};

void DNALandmarkBuilding::set_building_type(const string& building_type)
{
	m_building_type = building_type;
};

string DNALandmarkBuilding::get_building_type()
{
	return m_building_type;
};

void DNALandmarkBuilding::set_title(const string& title)
{
	m_title = title;
};

string DNALandmarkBuilding::get_title()
{
	return m_title;
};

void DNALandmarkBuilding::set_code(const string& code)
{
	m_code = code;
};

string DNALandmarkBuilding::get_code()
{
	return m_code;
};

void DNALandmarkBuilding::set_wall_color(LVector4f* color)
{
	m_wall_color = color;
};

LVector4f* DNALandmarkBuilding::get_wall_color()
{
	return m_wall_color;
};
